src/ptc/util/jms.clj

```diff
@@ -130,36 +121,83 @@
   [prefix workflow]
-  (let [result (str/join "/" [(cloud-prefix prefix workflow) "params.txt"])]
+  (let [cloud  (env-prefix prefix workflow)
+        result (str/join "/" [cloud "params.txt"])]
     (misc/gsutil "cp" "-" result :in (jms->params workflow))
     result))
 
-(defn jms->notification
-  "Push files to PREFIX and return notification for WORKFLOW."
+;; See https://broadinstitute.atlassian.net/wiki/spaces/GHConfluence/pages/2853961731/2021-07-28+AoU+Processing+Issue+Discussion
+;;
+(defn lookup-push-values
+  "Find ::push paths from WORKFLOW using PREFIX.
+   Use Green IDAT file as an example, the lookup order look like:
+   1. the value of the local :greenIDatPath key
+   2. (env-prefix    prefix workflow)/{CHIPWELL_BARCODE}_Grn.idat
+   3. (legacy-prefix prefix workflow)/{CHIPWELL_BARCODE}_Grn.idat
+
+   Return a single map of wfl-keys and their corresponding looked up
+   values.
+
+   Note we should not look at :cloudGreenIdatPath key in JMS as
+   it is for general arrays not AoU arrays!"
   [prefix workflow]
-  (let [cloud (cloud-prefix prefix workflow)
-        {:keys [::chip ::copy ::push]} (wfl-keys->jms-keys-for workflow)
-        chip-and-push (merge chip push)
-        sources (keep workflow (vals chip-and-push))]
-    (letfn [(rekey    [m [k v]] (assoc m k (v workflow)))
-            (cloudify [m [k v]]
-              (if-let [path (v workflow)]
-                (assoc m k (str/join "/" [cloud (last (str/split path #"/"))]))
-                m))]
-      (doseq [f sources]
-        (let [hash (misc/get-md5-hash f)]
-          (misc/gsutil "-h" (str "Content-MD5:" hash) "cp" f cloud)))
-      (reduce cloudify (reduce rekey {} copy) chip-and-push))))
+  (->> (for [[k local] (vec (::push wfl-keys->jms-keys))
+             :let [jms-local (local workflow)
+                   leaf      (last (str/split jms-local #"/"))
+                   glue-leaf #(str/join "/" [% leaf])
+                   lookup    (conj (concat
+                                     (map glue-leaf (all-versions-env-prefixes prefix workflow))
+                                     (map glue-leaf (all-versions-legacy-prefixes prefix workflow)))
+                               jms-local)]]
+         (let [object-to-us (loop [lookup lookup]
+                              (when (seq lookup)
+                                (if (misc/gcs-object-exists? (first lookup))
+                                  (first lookup)
+                                  (recur (rest lookup)))))]
+           ;; if failed to find a match, we report and throw the on-prem path for debug purposes
+           (if object-to-us
+             (log/info (format "Found a match at %s for %s." object-to-us leaf))
+             (do (log/errorf (format "Failed to find %s based on %s in %s!" leaf (first lookup) lookup))
+               (throw (Exception. (format "Failed to find %s based on %s in %s!" leaf (first lookup) lookup)))))
+           {k object-to-us}))
+    (into {})))
```

Comment:
object-to-usE?
Won't `conj` on the `concat`ed seqs add the local file to the head of the list.
Don't we want it at the end?
Will `misc/gcs-object-exists?` work on a local (NFS) file?